### DISCUSSION TITLE
Conditional Return type for RTK-Q Typescript

### DISCUSSION THREAD

**Discussion by dinoiam** [reduxjs/redux-toolkit#2052] [Q&A]
Hello, is there a way to specify different return types for a query? For example, If I have an endpoint that may return `Post` or `EnrichedPost`  can I somehow tell typescript which will be the return type based on the `postType` argument?
`useGetPostsQuery({ postType: 'normal' })` <-- In this case I would like currentData to be of type `Post`
`useGetPostsQuery({ postType: 'enriched' })` <-- In this case I would like currentData to be of type `EnrichedPost`. Thanks in advance.
```
const api = createApi({
  baseQuery: fetchBaseQuery({ baseUrl: '/ ' }),
  endpoints: (build) => ({
    getPosts: build.query<Post | EnrichedPost, { postType: 'normal' | 'enriched' }>({
      query: ({ postType }) => {
        return {
          url: `/posts`,
          body: {
            postType,
          },
        };
      },
    }),
  }),
})
```

**Comment by bryanprimus** [comment]
do you have updates on this? thinking of the same thing. and still can't figure this out

**Reply by phryneas** [reply]
It's unfortunately not possible - TypeScript has no language feature that would allow us to write an api supporting this. (It would need some kind of higher order generics)

**Reply by bryanprimus** [reply]
that said. does union type also not possible?

**Reply by phryneas** [reply]
You can put in a union type, but there will no connection between the "argument type" and the "result type". Those types go through multiple mapped types and never have any coupling.
